### ISSUE TITLE
[win10] Appium cannot find bundletool.jar in the PATH with node-which and where command

### ISSUE THREAD

**Issue by Archer456** [appium/appium#12269]
## The problem

appium-doctor 1.9.0  cant find my  local bundletool-all-0.8.0.jar 

## Environment
windows 10  os version 17763.348 
appium 1.11.1
appium-doctor 1.9.0 
nodejs 10.15.1
npm 6.4.1
android sdk 
![image](https://user-images.githubusercontent.com/18392901/53925048-81a96300-40b9-11e9-8674-4a9285baa5c4.png)

## Details

The hard one was the bundletool.jar - here are the steps I followed for that:

Searched for 'bundletool.jar' and found a place to download it from Google - I got 'bundletool-all-0.8.0.jar' from https://github.com/google/bundletool/releases

I renamed this file to just 'bundletool.jar' and placed in my Android SDK folder, under its own sub-folder called 'bundle-tools' - full path: D:\Android\android-sdk\bundle-tools\

![image](https://user-images.githubusercontent.com/18392901/53924798-68ec7d80-40b8-11e9-8f4b-da117dcef0b6.png)

Added the 'bundle-tools' folder to Path with the following addition in my PATH
![image](https://user-images.githubusercontent.com/18392901/53924944-165f9100-40b9-11e9-910b-83b1b2fecc11.png)

Now running which 'bundletool.jar' is happy and finds it in the 'bundle-tools' folder I created (see step 2)

Running 'appium-doctor' is now satisfied - and doesn't report any missing dependencies

## Link to Appium logs
`C:\Users\cmd>appium-doctor
info AppiumDoctor Appium Doctor v.1.9.0
info AppiumDoctor ### Diagnostic for necessary dependencies starting ###
info AppiumDoctor  ✔ The Node.js binary was found at: D:\nodejs\node.EXE
info AppiumDoctor  ✔ Node version is 10.15.1
info AppiumDoctor  ✔ ANDROID_HOME is set to: D:\Android\android-sdk
info AppiumDoctor  ✔ JAVA_HOME is set to: C:\Program Files\Java\jdk1.8.0_201
info AppiumDoctor  ✔ adb exists at: D:\Android\android-sdk\platform-tools\adb.exe
info AppiumDoctor  ✔ android exists at: D:\Android\android-sdk\tools\android.bat
info AppiumDoctor  ✔ emulator exists at: D:\Android\android-sdk\tools\emulator.exe
info AppiumDoctor  ✔ Bin directory of %JAVA_HOME% is set
info AppiumDoctor ### Diagnostic for necessary dependencies completed, no fix needed. ###
info AppiumDoctor
info AppiumDoctor ### Diagnostic for optional dependencies starting ###
info AppiumDoctor  ✔ Python required by node-gyp (used by heapdump) is installed at: D:\Python27\python.EXE. Installed version is: 2.7.14
WARN AppiumDoctor  ✖ opencv4nodejs cannot be found.
WARN AppiumDoctor  ✖ ffmpeg cannot be found
WARN AppiumDoctor  ✖ bundletool.jar cannot be found
info AppiumDoctor ### Diagnostic for optional dependencies completed, 3 fixes possible. ###
info AppiumDoctor
info AppiumDoctor ### Optional Manual Fixes ###
info AppiumDoctor The configuration can install optionally. Please do the following manually:
WARN AppiumDoctor  ➜ Why opencv4nodejs is needed and how to install it: https://github.com/appium/appium/blob/master/docs/en/writing-running-appium/image-comparison.md
WARN AppiumDoctor  ➜ ffmpeg is needed to record screen features. Please read https://www.ffmpeg.org/ to install it
WARN AppiumDoctor  ➜ bundletool.jar is used to handle Android App Bundle. Please read http://appium.io/docs/en/writing-running-appium/android/android-appbundle/ to install it
info AppiumDoctor
info AppiumDoctor ###
info AppiumDoctor
info AppiumDoctor Bye! Run appium-doctor again when all manual fixes have been applied!
info AppiumDoctor`

I can use bundletool like this 

`D:\Android\android-sdk\bundle-tools>java -jar bundletool-all-0.8.0.jar --help
Error: You have to specify a command.
Synopsis: bundletool <command> ...

Use 'bundletool help <command>' to learn more about the given command.

build-bundle command:
    Builds an Android App Bundle from a set of Bundle modules provided as zip
    files.

build-apks command:
    Generates an APK Set archive containing either all possible split APKs and
    standalone APKs or APKs optimized for the connected device (see connected-
    device flag).

extract-apks command:
    Extracts from an APK Set the APKs that should be installed on a given
    device.

get-device-spec command:
    Writes out a JSON file containing the device specifications (i.e. features
    and properties) of the connected Android device.

install-apks command:
    Installs APKs extracted from an APK Set to a connected device. Replaces
    already installed package.

validate command:
    Verifies the given Android App Bundle is valid and prints out information
    about it.

dump command:
    Prints files or extract values from the bundle in a human-readable form.

get-size command:
    Computes the min and max download sizes of APKs served to different devices
    configurations from an APK Set.

version command:
    Prints the version of BundleTool.

D:\Android\android-sdk\bundle-tools>bunletool help
'bunletool' 不是内部或外部命令，也不是可运行的程序
或批处理文件。`

## Code To Reproduce Issue [ Good To Have ]

Please remember that with sample code it's easier to reproduce the bug and it's much faster to fix it.

Please git clone https://github.com/appium/appium and from the `sample-code` directory, use one of your favourite languages and sample apps to reproduce the issue.

In case a similar scenario is missing in sample-code, please submit a PR with one of the sample apps provided.

**Comment by mykola-mokhnach** [comment]
Make sure Appium Doctor app properly gets all environment variables (especially the PATH one)

**Comment by KazuCocoa** [comment]
What is the output of PATH in your env?
e.g. https://superuser.com/questions/1216658/path-environment-variable-windows-10-echo-path-on-command-prompt-shows-only

**Comment by Archer456** [comment]
> What is the output of PATH in your env?
> e.g. https://superuser.com/questions/1216658/path-environment-variable-windows-10-echo-path-on-command-prompt-shows-only

![image](https://user-images.githubusercontent.com/18392901/54094420-cbf15380-43db-11e9-860d-68d56aef1b4a.png)

**Comment by KazuCocoa** [comment]
Can you invoke the `bundletool` as `D:\Android\android-sdk\bundle-tools\bundletool.jar`?

**Comment by Archer456** [comment]
> Can you invoke the `bundletool` as `D:\Android\android-sdk\bundle-tools\bundletool.jar`?
like this I cat use
![image](https://user-images.githubusercontent.com/18392901/54094850-af0a4f80-43de-11e9-974f-cdb7213a52cf.png)

**Comment by KazuCocoa** [comment]
As addressed in http://appium.io/docs/en/writing-running-appium/android/android-appbundle/ which is included in the doctor's message, the bundler jar name should be `bundletool.jar`
> Appium looks for bundletool.jar in your local environemnt

Could you try out again after renaming the jar?

**Comment by Archer456** [comment]
> As addressed in http://appium.io/docs/en/writing-running-appium/android/android-appbundle/ which is included in the doctor's message, the bundler jar name should be `bundletool.jar`
> 
> > Appium looks for bundletool.jar in your local environemnt
> 
> Could you try out again after renaming the jar?

![image](https://user-images.githubusercontent.com/18392901/54095165-b92d4d80-43e0-11e9-834e-2187d68b3e45.png)

**Comment by KazuCocoa** [comment]
So, you could call `D:\Android\android-sdk\bundle-tools\bundletool.jar` (Not `bundletool-all-0.8.0.jar`)), but Appium doctor did not find the `bundletool.jar`, right?

**Comment by Archer456** [comment]
> So, you could call `D:\Android\android-sdk\bundle-tools\bundletool.jar` (Not `bundletool-all-0.8.0.jar`)), but Appium doctor did not find the `bundletool.jar`, right?

right !

`C:\Users\cmd>java -jar D:\Android\android-sdk\bundle-tools\bundletool.jar --help
Error: You have to specify a command.
Synopsis: bundletool <command> ...

Use 'bundletool help <command>' to learn more about the given command.

build-bundle command:
    Builds an Android App Bundle from a set of Bundle modules provided as zip
    files.

build-apks command:
    Generates an APK Set archive containing either all possible split APKs and
    standalone APKs or APKs optimized for the connected device (see connected-
    device flag).

extract-apks command:
    Extracts from an APK Set the APKs that should be installed on a given
    device.

get-device-spec command:
    Writes out a JSON file containing the device specifications (i.e. features
    and properties) of the connected Android device.

install-apks command:
    Installs APKs extracted from an APK Set to a connected device. Replaces
    already installed package.

validate command:
    Verifies the given Android App Bundle is valid and prints out information
    about it.

dump command:
    Prints files or extract values from the bundle in a human-readable form.

get-size command:
    Computes the min and max download sizes of APKs served to different devices
    configurations from an APK Set.

version command:
    Prints the version of BundleTool.


C:\Users\cmd>bundletool.jar --help

C:\Users\cmd>java -jar bundletool.jar --help
Error: Unable to access jarfile bundletool.jar

C:\Users\cmd>d:

D:\>cd Android\android-sdk\bundle-tools

D:\Android\android-sdk\bundle-tools>dir
 驱动器 D 中的卷是 本地磁盘
 卷的序列号是 AC5A-320E

 D:\Android\android-sdk\bundle-tools 的目录

2019/03/11  09:31    <DIR>          .
2019/03/11  09:31    <DIR>          ..
2019/03/06  17:52        38,230,949 bundletool.jar
               1 个文件     38,230,949 字节
               2 个目录 119,541,432,320 可用字节

D:\Android\android-sdk\bundle-tools>java -jar bundletool.jar --help
Error: You have to specify a command.
Synopsis: bundletool <command> ...

Use 'bundletool help <command>' to learn more about the given command.

build-bundle command:
    Builds an Android App Bundle from a set of Bundle modules provided as zip
    files.

build-apks command:
    Generates an APK Set archive containing either all possible split APKs and
    standalone APKs or APKs optimized for the connected device (see connected-
    device flag).

extract-apks command:
    Extracts from an APK Set the APKs that should be installed on a given
    device.

get-device-spec command:
    Writes out a JSON file containing the device specifications (i.e. features
    and properties) of the connected Android device.

install-apks command:
    Installs APKs extracted from an APK Set to a connected device. Replaces
    already installed package.

validate command:
    Verifies the given Android App Bundle is valid and prints out information
    about it.

dump command:
    Prints files or extract values from the bundle in a human-readable form.

get-size command:
    Computes the min and max download sizes of APKs served to different devices
    configurations from an APK Set.

version command:
    Prints the version of BundleTool.`

**Comment by KazuCocoa** [comment]
Do you have an npm environment?
If you have, could you follow below steps and tell me the output of `which.sync`?

```
git clone git@github.com:appium/appium-desktop.git
cd appium-desktop
npm install
# Finish

node # Go to node console
> var which = require('which')
> which.sync('bundletool.jar')
```

Could you try out below as well?
```
# In powershell (Not in the node console)
where bundletool.jar
```

I wonder we cannot get a path to `bundletool.jar` in windows env by node `which` module.
(If so, we should update the logic in appium as well.)

**Comment by Archer456** [comment]
> Do you have an npm environment?
> If you have, could you follow below steps and tell me the output of `which.sync`?
> 
> ```
> git clone git@github.com:appium/appium-desktop.git
> cd appium-desktop
> npm install
> # Finish
> 
> node # Go to node console
> > var which = require('which')
> > which.sync('bundletool.jar')
> ```
> Could you try out below as well?
> 
> ```
> # In powershell (Not in the node console)
> where bundletool.jar
> ```
> I wonder we cannot get a path to `bundletool.jar` in windows env by node `which` module.
> (If so, we should update the logic in appium as well.)


I do like this:
1  `D:\workspace\git>git clone git@github.com:appium/appium-desktop.git
Cloning into 'appium-desktop'...
Warning: Permanently added the RSA host key for IP address '13.229.188.59' to the list of known hosts.
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
# i home no  access rights
# i think  you want me to git clone this?
D:\workspace\git>git clone https://github.com/appium/appium-desktop.git
Cloning into 'appium-desktop'...
remote: Enumerating objects: 343, done.
remote: Counting objects: 100% (343/343), done.
remote: Compressing objects: 100% (144/144), done.
remote: Total 6803 (delta 240), reused 267 (delta 185), pack-reused 6460 eceiving objects:  99% (6735/6803), 7.02 MiB | 405.00 KiB/s 
Receiving objects: 100% (6803/6803), 7.05 MiB | 399.00 KiB/s, done.
Resolving deltas: 100% (4689/4689), done.`

# but npm install error
`D:\workspace\git\appium-desktop>npm install
npm WARN tarball tarball data for antd@3.11.2 (sha512-bwc9L+5Dk6sGbtA2Nr9rnfchM5JJ28Zb/AZTzMzevwA/89CFm5Eq4uBhFtYrITv8R2+p0FXyPo4lDVUQ16IlPQ==) seems to be corrupted. Trying one more time.
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\@sinonjs\samsam-70ec64ed\package.json'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\@sinonjs\samsam-70ec64ed\LICENSE'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\@sinonjs\samsam-70ec64ed\README.md'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\fs-extra-3469debf\lib\util\assign.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\rc-calendar-d273c3d0\lib\month\MonthPanel.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\fs-extra-3469debf\lib\util\buffer.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\app-builder-lib-1d1f69c3\templates\appx\appxmanifest.xml'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\external-editor-23a451d5\main\errors\CreateFileError.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\fs-extra-3469debf\lib\util\utimes.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\rc-calendar-d273c3d0\lib\month\MonthTable.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\app-builder-lib-1d1f69c3\templates\appx\priconfig.xml'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\external-editor-23a451d5\main\errors\LaunchEditorError.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\debug-dd5b655b\dist\debug.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\less-da7e1eec\test\less-bom\sourcemaps\basic.json'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\external-editor-23a451d5\main\errors\ReadFileError.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\less-da7e1eec\test\less-bom\sourcemaps\basic.less'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\external-editor-23a451d5\main\errors\RemoveFileError.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\less-da7e1eec\test\less-bom\sourcemaps\imported.css'
npm WARN tar ENOENT: no such file or directory, lstat 'D:\workspace\git\appium-desktop\node_modules\.staging\@babel\plugin-proposal-unicode-property-regex-ec6c243a\test\fixtures\with-unicode-flag'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\less-da7e1eec\test\less-bom\static-urls\urls.less'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\inquirer-7ba7a007\lib\utils\events.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\inquirer-7ba7a007\lib\utils\paginator.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\inquirer-7ba7a007\lib\utils\readline.js'
npm WARN tar ENOENT: no such file or directory, lstat 'D:\workspace\git\appium-desktop\node_modules\.staging\less-da7e1eec\test\less-bom\strict-units'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\inquirer-7ba7a007\lib\utils\screen-manager.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\less-da7e1eec\test\less-bom\strings.less'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\inquirer-7ba7a007\lib\utils\utils.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\pako-69f74aad\lib\utils\common.js'
npm WARN tar ENOENT: no such file or directory, chmod 'D:\workspace\git\appium-desktop\node_modules\.staging\@babel\plugin-proposal-unicode-property-regex-ec6c243a'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\pako-69f74aad\lib\utils\strings.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\rc-calendar-d273c3d0\lib\year\YearPanel.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\appium-base-driver-cd921e8d\test\basedriver\fixtures\.DS_Store'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\appium-base-driver-cd921e8d\test\basedriver\fixtures\BadZippedApp.zip'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\appium-base-driver-cd921e8d\test\basedriver\fixtures\custom-element-finder-bad.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\appium-base-driver-cd921e8d\test\basedriver\fixtures\custom-element-finder.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\appium-base-driver-cd921e8d\test\basedriver\fixtures\FakeAndroidApp.apk'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\appium-base-driver-cd921e8d\test\basedriver\fixtures\FakeAndroidApp.asd'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\appium-base-driver-cd921e8d\test\basedriver\fixtures\FakeIOSApp.app'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\appium-base-driver-cd921e8d\test\basedriver\fixtures\FakeIOSApp.app.zip'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\appium-base-driver-cd921e8d\test\basedriver\fixtures\FakeIOSApp.ipa'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\jsdom-88b5d6b1\lib\jsdom\named-properties-tracker.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\appium-base-driver-cd921e8d\test\basedriver\helpers-e2e-specs.js'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\jsdom-88b5d6b1\lib\jsdom\utils.js'
npm WARN tar ENOENT: no such file or directory, lstat 'D:\workspace\git\appium-desktop\node_modules\.staging\app-builder-lib-1d1f69c3\templates\icons\proton-native'
npm WARN tar ENOENT: no such file or directory, lstat 'D:\workspace\git\appium-desktop\node_modules\.staging\jimp-fa21a061\fonts\open-sans\open-sans-10-black'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\ajv-f0098d3b\lib\refs\$data.json'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\ajv-f0098d3b\lib\refs\json-schema-draft-04.json'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\ajv-f0098d3b\lib\refs\json-schema-draft-06.json'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\ajv-f0098d3b\lib\refs\json-schema-v5.json'
npm WARN appium-desktop@1.11.1 license should be a valid SPDX license expression

npm ERR! path D:\workspace\git\appium-desktop\node_modules\.staging\antd-4b364ce2\dist\antd-with-locales.js
npm ERR! code EPERM
npm ERR! errno -4048
npm ERR! syscall unlink
npm ERR! Error: EPERM: operation not permitted, unlink 'D:\workspace\git\appium-desktop\node_modules\.staging\antd-4b364ce2\dist\antd-with-locales.js'
npm ERR!  { [Error: EPERM: operation not permitted, unlink 'D:\workspace\git\appium-desktop\node_modules\.staging\antd-4b364ce2\dist\antd-with-locales.js']
npm ERR!   cause:
npm ERR!    { Error: EPERM: operation not permitted, unlink 'D:\workspace\git\appium-desktop\node_modules\.staging\antd-4b364ce2\dist\antd-with-locales.js'
npm ERR!      errno: -4048,
npm ERR!      code: 'EPERM',
npm ERR!      syscall: 'unlink',
npm ERR!      path:
npm ERR!       'D:\\workspace\\git\\appium-desktop\\node_modules\\.staging\\antd-4b364ce2\\dist\\antd-with-locales.js' },
npm ERR!   stack:
npm ERR!    'Error: EPERM: operation not permitted, unlink \'D:\\workspace\\git\\appium-desktop\\node_modules\\.staging\\antd-4b364ce2\\dist\\antd-with-locales.js\'',
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'unlink',
npm ERR!   path:
npm ERR!    'D:\\workspace\\git\\appium-desktop\\node_modules\\.staging\\antd-4b364ce2\\dist\\antd-with-locales.js',
npm ERR!   parent: 'appium-desktop' }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator (though this is not recommended).
npm WARN tar ENOENT: no such file or directory, lstat 'D:\workspace\git\appium-desktop\node_modules\.staging\eslint-a8c1b644\lib\util'
npm WARN tar ENOENT: no such file or directory, open 'D:\workspace\git\appium-desktop\node_modules\.staging\eslint-a8c1b644\lib\util\report-translator.js'

npm ERR! A complete log of this run can be found in:
npm ERR!     D:\nodejs\node-cache\_logs\2019-03-11T02_33_35_881Z-debug.log`

2 
![image](https://user-images.githubusercontent.com/18392901/54096929-ed0c7100-43e8-11e9-9b42-37054cdd77b5.png)
not find bundletool.jar

**Comment by Archer456** [comment]
`Windows PowerShell
版权所有 (C) Microsoft Corporation。保留所有权利。

PS C:\Users\cmd> node -v
v10.15.1
PS C:\Users\cmd> node
> var which = require('which')
{ Error: Cannot find module 'which'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:581:15)
    at Function.Module._load (internal/modules/cjs/loader.js:507:25)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18) code: 'MODULE_NOT_FOUND' }
>
`

**Comment by KazuCocoa** [comment]
mm, what about like below as _1_? Install `which` as global scope and call it via CLI.

```
npm install -g which
/Users/kazu/.nodebrew/node/v11.9.0/bin/which -> /Users/kazu/.nodebrew/node/v11.9.0/lib/node_modules/which/bin/which
+ which@1.3.1

/Users/kazu/.nodebrew/node/v11.9.0/bin/which bundletool.jar
```

Thanks for _2_

**Comment by Archer456** [comment]
> mm, what about like below as _1_? Install `which` as global scope and call it via CLI.
> 
> ```
> npm install -g which
> /Users/kazu/.nodebrew/node/v11.9.0/bin/which -> /Users/kazu/.nodebrew/node/v11.9.0/lib/node_modules/which/bin/which
> + which@1.3.1
> 
> /Users/kazu/.nodebrew/node/v11.9.0/bin/which bundletool.jar
> ```
> Thanks for _2_

`C:\Users\cmd>npm install -g which
D:\nodejs\node-global\which -> D:\nodejs\node-global\node_modules\which\bin\which
+ which@1.3.1
added 2 packages from 1 contributor in 0.961s

C:\Users\cmd>node
> var which = require('which')
{ Error: Cannot find module 'which'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:581:15)
    at Function.Module._load (internal/modules/cjs/loader.js:507:25)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18) code: 'MODULE_NOT_FOUND' }
>

`

**Comment by KazuCocoa** [comment]
You can run `D:\nodejs\node-global\which` directly in this case. You do not need to call on node console.

So, maybe like `D:\nodejs\node-global\which bundletool.jar`

**Comment by Archer456** [comment]
> You can run `D:\nodejs\node-global\which` directly in this case. You do not need to call on node console.
> 
> So, maybe like `D:\nodejs\node-global\which bundletool.jar`

![image](https://user-images.githubusercontent.com/18392901/54097915-41195480-43ed-11e9-9514-0d10a42ec175.png)
no output message

**Comment by Archer456** [comment]
@KazuCocoa  I guess your operating system is mac

So you don't have a system environment to verify this.


Appium is very easy to install and use under Mac and ubuntu.

But there are always many strange problems under Windows.

I recommend that testing be strengthened in three systems environments before release

**Comment by KazuCocoa** [comment]
Yes, since I need iOS dev env.
In some repositories start using Azure as CI. Ruby clients run tests on windows env, too. This kind of CI will help us to run tests on windows, I hope.

That's right, I should buy a good Windows PC first tho..
(I hope strong windows users join our development)

**Comment by Archer456** [comment]
> Yes, since I need iOS dev env.
> In some repositories start using Azure as CI. Ruby clients run tests on windows env, too. This kind of CI will help us to run tests on windows, I hope.
> 
> That's right, I should buy a good Windows PC first tho..
> (I hope strong windows users join our development)

Thank you very much for your answer.

Expect the project team to strengthen pre-release testing in different system environments

I also wish you to use Windows 10 to test this problem as soon as possible.

**Comment by ishubhamjain** [comment]
Facing the exactly same issue.

**Comment by duoniduoni** [comment]
Facing the same issue on ubuntu 14.04

**Comment by KazuCocoa** [comment]
What about adding executable permission to the `bundletool.jar` like `$ chmod 655 xxx/bundletool.jar`?

**Comment by duoniduoni** [comment]
> What about adding executable permission to the `bundletool.jar` like `$ chmod 655 xxx/bundletool.jar`?

it works , thanks ~

**Comment by astraydog** [comment]
@Archer456 , I am also facing the same problem on windows 10

**Comment by Archer456** [comment]
> What about adding executable permission to the `bundletool.jar` like `$ chmod 655 xxx/bundletool.jar`?

yes in my ubuntu18.04.1   it is OK

like this:
1 put bundletool.jar in /Android/Sdk/bundle-tools
   sudo chmod 755 bundletool.jar
  sudo chown cmd:cmd bundletool.jar
2 edit my ~/.profile     my path is all here
 `#Android-sdk-linux Path
export ANDROID_HOME=$HOME/Android/Sdk
export PATH=$PATH::$ANDROID_HOME/tools/:$ANDROID_HOME/platform-tools/:$ANDROID_HOME/bundle-tools/
`

3 `cmd@TR:~$ appium -v
1.11.1
cmd@TR:~$ appium-doctor
info AppiumDoctor Appium Doctor v.1.9.0
info AppiumDoctor ### Diagnostic for necessary dependencies starting ###
info AppiumDoctor  ✔ The Node.js binary was found at: /opt/nodejs/bin/node
info AppiumDoctor  ✔ Node version is 10.15.1
info AppiumDoctor  ✔ ANDROID_HOME is set to: /home/cmd/Android/Sdk
info AppiumDoctor  ✔ JAVA_HOME is set to: /usr/lib/jvm/java-8-oracle
info AppiumDoctor  ✔ adb exists at: /home/cmd/Android/Sdk/platform-tools/adb
info AppiumDoctor  ✔ android exists at: /home/cmd/Android/Sdk/tools/android
info AppiumDoctor  ✔ emulator exists at: /home/cmd/Android/Sdk/tools/emulator
info AppiumDoctor  ✔ Bin directory of $JAVA_HOME is set
info AppiumDoctor ### Diagnostic for necessary dependencies completed, no fix needed. ###
info AppiumDoctor 
info AppiumDoctor ### Diagnostic for optional dependencies starting ###
info AppiumDoctor  ✔ Python required by node-gyp (used by heapdump) is installed at: /home/cmd/.pyenv/shims/python. Installed version is: 2.7.15
WARN AppiumDoctor  ✖ opencv4nodejs cannot be found.
WARN AppiumDoctor  ✖ ffmpeg cannot be found
info AppiumDoctor  ✔ bundletool.jar is installed at: /home/cmd/Android/Sdk/bundle-tools/bundletool.jar
info AppiumDoctor ### Diagnostic for optional dependencies completed, 2 fixes possible. ###
info AppiumDoctor 
info AppiumDoctor ### Optional Manual Fixes ###
info AppiumDoctor The configuration can install optionally. Please do the following manually:
WARN AppiumDoctor  ➜ Why opencv4nodejs is needed and how to install it: https://github.com/appium/appium/blob/master/docs/en/writing-running-appium/image-comparison.md
WARN AppiumDoctor  ➜ ffmpeg is needed to record screen features. Please read https://www.ffmpeg.org/ to install it
info AppiumDoctor 
info AppiumDoctor ###
info AppiumDoctor 
info AppiumDoctor Bye! Run appium-doctor again when all manual fixes have been applied!
`

**Comment by portlander** [comment]
> @KazuCocoa I guess your operating system is mac
> 
> So you don't have a system environment to verify this.
> 
> Appium is very easy to install and use under Mac and ubuntu.
> 
> But there are always many strange problems under Windows.
> 
> I recommend that testing be strengthened in three systems environments before release

After adding .JAR to PATHEXT , my appium-doctor can then report bundletool.jar installed.
Also, which bundletool.jar can find it too.

**Comment by mykola-mokhnach** [comment]
@KazuCocoa Is this issue still relevant?

**Comment by kashban** [comment]
> 
> After adding .JAR to PATHEXT , my appium-doctor can then report bundletool.jar installed.
> Also, node-which bundletool.jar can find it too.

Thanks, that did the trick for me. Adding it to the PATH wasn't enough.

**Comment by KazuCocoa** [comment]
In doctor 1.10.0 has a message to address https://github.com/appium/appium/issues/12269#issuecomment-482483574 in windows environment.

**Comment by Archer456** [comment]
> the doctor 1.10.0 has a message to address [#12269 (comment)](https://github.com/appium/appium/issues/12269#issuecomment-482483574) in windows environment.

Thank you very much . I've solved this after i update appium-doctor to 1.10.0

We have to add system environment variables manually. 

just like this 
![image](https://user-images.githubusercontent.com/18392901/57186686-56ef5580-6f16-11e9-8224-cc5f60552212.png)

and PATH like this:
![image](https://user-images.githubusercontent.com/18392901/57186698-81d9a980-6f16-11e9-9ba7-19e82e680229.png)

Then you'll find bundletool. jar

![image](https://user-images.githubusercontent.com/18392901/57186663-ffe98080-6f15-11e9-81bb-6fbc7b92dc28.png)

**Comment by sicute** [comment]
i use mac and only move from Download folder to path /Users/username/Library/Android/sdk/build-tools

>
$ mv Downloads/bundletool-all-0.10.1.jar /Users/username/Library/Android/sdk/build-tools/bundletool.jar 
$ chmod 655 /Users/username/Library/Android/sdk/build-tools/bundletool.jar

**Comment by kapoorlakshya** [comment]
@sicute Thanks for sharing your steps. This worked for me on my Mac. `build-tools` seems to be an appropriate place for this file.

**Comment by emiryazicii** [comment]
> > the doctor 1.10.0 has a message to address [#12269 (comment)](https://github.com/appium/appium/issues/12269#issuecomment-482483574) in windows environment.
> 
> Thank you very much . I've solved this after i update appium-doctor to 1.10.0
> 
> We have to add system environment variables manually.
> 
> just like this ![image](https://user-images.githubusercontent.com/18392901/57186686-56ef5580-6f16-11e9-8224-cc5f60552212.png)
> 
> and PATH like this: ![image](https://user-images.githubusercontent.com/18392901/57186698-81d9a980-6f16-11e9-9ba7-19e82e680229.png)
> 
> Then you'll find bundletool. jar
> 
> ![image](https://user-images.githubusercontent.com/18392901/57186663-ffe98080-6f15-11e9-81bb-6fbc7b92dc28.png)

That also worked for me , thanks a lot for sharing.

**Comment by mdogancay** [comment]
> > @KazuCocoa I guess your operating system is mac
> > So you don't have a system environment to verify this.
> > Appium is very easy to install and use under Mac and ubuntu.
> > But there are always many strange problems under Windows.
> > I recommend that testing be strengthened in three systems environments before release
> 
> After adding .JAR to PATHEXT , my appium-doctor can then report bundletool.jar installed. Also, node-which bundletool.jar can find it too.

it's works on windows 10(11?). PATHEXT is a system environment variable of windows 10(11?). edit from advanced system settings->environment variables->system variables and add ;.JAR to it's end.
